[clang][SPIRV] Add builtin for OpGenericCastToPtrExplicit and its SPIR-V friendly binding

Naghasan opened the pull request:
The patch introduce __builtin_spirv_generic_cast_to_ptr_explicit which is lowered to the llvm.spv.generic.cast.to.ptr.explicit intrinsic.

The SPIR-V builtins are now split into 3 differents file: BuiltinsSPIRVCore.td,
BuiltinsSPIRVVK.td for Vulkan specific builtins, BuiltinsSPIRVCL.td for OpenCL specific builtins
and BuiltinsSPIRVCommon.td for common ones.

The patch also introduces a new header defining its SPIR-V friendly equivalent (__spirv_GenericCastToPtrExplicit_ToGlobal, __spirv_GenericCastToPtrExplicit_ToLocal and __spirv_GenericCastToPtrExplicit_ToPrivate). The functions are declared as aliases to the new builtin allowing C-like languages to have a definition to rely on as well as gaining proper front-end diagnostics.

The motivation for the header is to provide a stable binding for applications or library (such as SYCL) and allows non SPIR-V targets to provide an implementation (via libclc or similar to how it is done for gpuintrin.h).

Commit:
[clang][SPIRV] Add builtin for OpGenericCastToPtrExplicit and its SPIR-V friendly binding

The patch introduce __builtin_spirv_generic_cast_to_ptr_explicit which is lowered to
the llvm.spv.generic.cast.to.ptr.explicit intrinsic.

The patch also introduces a new header defining its SPIR-V friendly equivalent
(__spirv_GenericCastToPtrExplicit_ToGlobal, __spirv_GenericCastToPtrExplicit_ToLocal
and __spirv_GenericCastToPtrExplicit_ToPrivate). The functions are declared as aliases
to the new builtin allowing C-like languages to have a definition to rely on as well as
gaining proper front-end diagnostics.

Naghasan commented:
Sorry I can't assign reviewers, @JonChesterfield  @jhuber6 as you are involved in gpuintrin.h, I'd welcome feedbacks here (feel free to ping other relevant persons)

For the SPIR-V side of thing @VyacheslavLevytskyy @farzonl I appreciate your feedbacks as well :)


FYI @tahonermann

farzonl commented on clang/lib/Basic/Targets/SPIR.cpp:
```diff
@@ -35,8 +35,10 @@ static constexpr Builtin::Info BuiltinInfos[] = {
-  return {{&BuiltinStrings, BuiltinInfos}};
+BaseSPIRTargetInfo::getTargetBuiltins() const {
+  if (getTriple().isSPIRV())
+    return {{&BuiltinStrings, BuiltinInfos}};
```
I don't understand why this change (in SPIR.cpp and SPIR.h) is necessesary.  The existing code accomplished the same thing with method overriding. The original code also had some defensive programming safety to me. We went  from a compile time enforcement of  SPIRVTargetInfo== BuiltinsSPIRV.td to now a runtime enforcement. The problem with that is the next guy that comes along and makes a change might assume the builtins so far are SPIRV and thats why you are checking the triple, then add a new SPIR flavor builtin that would not be valid for SPIRV.

Naghasan replied:
Oh, that's a left over of my initial prototype in intel/llvm sorry. For historical reasons `spir` / `spir64` is also used there, so had to hack a bit to allow proper testing there. The change should be moving it to `BaseSPIRVTargetInfo` to only enable `spirv`, `spirv32` and `spirv64`. I updated the PR.

farzonl replied:
Ok we might have a hierarchy problem then how do we say a builtin is valid for `spirv` but not `spirv32` or `spriv64` and vice versa. see https://github.com/llvm/llvm-project/pull/137805#discussion_r2066520947

Naghasan replied:
I had a few thought about this, but there is a bit more to it:

1. Vulkan vs OpenCL environment (and for which target version)
2. Enabled vs disabled extensions
3. SPIR-V version

each will have an impact on the set of enabled builtins. A split between common, Vk env (`spirv`) and OpenCL env (`spirv32` / `spirv64`) builtins is probably a good step towards answering (1) by construction. What I don't like though is you'll end up with an "unknown builtin" rather than a clear message.
For the rest, I wanted to create an RFC to suggest the use of target features as there is plenty of existing infrastructure for this job.

farzonl replied:
What if we had a `BuiltinsSPIRVCore.td` `BuiltinsSPIRVVK.td` and a `BuiltinsSPIRVCL.td` The stuff shared between CL and VK could live in Core and the VK and CL builtins will include the Core one. That should let us have seperate `SPIRVTargetInfo` and `SPIRV64TargetInfo`/`SPIRV32TargetInfo`. You won't have an unknown builtin if we do it this way because we would never expose the builtins that aren't valid for the target.

farzonl replied:
I'll consider this issue resolved though if you plan to address it via an RFC.

Naghasan replied:
> What if we had a BuiltinsSPIRVCore.td BuiltinsSPIRVVK.td and a BuiltinsSPIRVCL.td

I was looking into doing this.

> You won't have an unknown builtin if we do it this way because we would never expose the builtins that aren't valid for the target.

You will if you use a VK one when targeting `spirv64` rather than a clear message. But it is nitpicking and I only have a weak preference on the matter.

Naghasan replied:
@farzonl Apply the suggestion, although you can't really split into 2 distinct set cleanly (common + vk and common + cl) due to the ID assignment (part is done dynamically). So I went for a unique set and diagnose when vk builtins are used with spirv32 or spirv32 or when cl builtins are used with spirv.

farzonl commented on clang/lib/Basic/Targets/SPIR.cpp:
```diff
@@ -35,8 +35,10 @@ static constexpr Builtin::Info BuiltinInfos[] = {
-SPIRVTargetInfo::getTargetBuiltins() const {
-  return {{&BuiltinStrings, BuiltinInfos}};
+BaseSPIRTargetInfo::getTargetBuiltins() const {
+  if (getTriple().isSPIRV())
```
if you are using this to extend  the builtins to spirv32/spirv64 I have concerns that not all builtins will be valid for those targets. For example `__builtin_spirv_reflect` is only valid when targeting spirv because thats the target that enables GLSL extentions. It is not valid for spirv32/64 because there is no equivalent opencl extention. To expose across all targets without exposing builtins that break could be tricky and might require special codegen to emulate the behavior of other targets. It might be simpler to just do a new tablegen file for spirv32/spriv64.

farzonl commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -5837,12 +5838,13 @@ static void handleBuiltinAliasAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
        !S.ARM().CdeAliasValid(BuiltinID, AliasName)) ||
       (IsRISCV && !S.RISCV().isAliasValid(BuiltinID, AliasName)) ||
-      (!IsAArch64 && !IsARM && !IsRISCV && !IsHLSL)) {
+      (!IsAArch64 && !IsARM && !IsRISCV && !IsHLSL && !IsSPIRV)) {
```
This shouldn't impact HLSL's ability to target SPIRV will it? More specifically will this impact HLSL's use of SPIRV builtins?

Naghasan replied:
I'm not affecting HLSL in any way here.

Commit:
Address feedback: split common, vk and ocl builtins

The SPIR-V builtins are now split into 3 differents file: BuiltinsSPIRVCore.td,
BuiltinsSPIRVVK.td for Vulkan specific builtins, BuiltinsSPIRVCL.td for OpenCL specific builtins
and BuiltinsSPIRVCommon.td for common ones.

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -97,10 +97,10 @@ static Value *EmitTargetArchBuiltinExpr(CodeGenFunction *CGF,
   case llvm::Triple::riscv64:
     return CGF->EmitRISCVBuiltinExpr(BuiltinID, E, ReturnValue);
   case llvm::Triple::spirv:
```
The change makes it so the spirv target can call `EmitAMDGPUBuiltinExpr` if the triple has AMDHSA. Is that an intended change?

Naghasan replied:
nope, I fixed it

farzonl commented on clang/lib/Sema/SemaSPIRV.cpp:
```diff
@@ -33,8 +46,114 @@ static bool CheckAllArgsHaveSameType(Sema *S, CallExpr *TheCall) {
+                                              unsigned BuiltinID,
                                               CallExpr *TheCall) {
+  if (BuiltinID >= SPIRV::FirstVKBuiltin && BuiltinID <= SPIRV::LastVKBuiltin)
+    if (TI.getTriple().getArch() != llvm::Triple::spirv) {
```
Can we make this one conditional? same with like 150-151.

Naghasan replied:
done

VyacheslavLevytskyy reviewed (approved):
LGTM

farzonl reviewed (approved):
LGTM

Naghasan commented:
Just opened a PR for this https://github.com/llvm/llvm-project/pull/141954
